src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -158,6 +159,14 @@ export class WorkspaceSavedObjectsClientWrapper {
     objectPermissionModes: WorkspacePermissionMode[],
     validateAllWorkspaces = true
   ) {
+    /**
+     * A data source saved object without workspaces attributes will be treated as a global data source.
+     * This kind of data source is not allowed for non dashboard admin. The dashboard admin will bypass this
+     * client wrapper, so denied all access to the global data source saved object here.
+     */
+    if (savedObject.type === DATA_SOURCE_SAVED_OBJECT_TYPE && !savedObject.workspaces) {
```

Comment:
Now I feel it does make sense to check if `workspaces` is empty
```suggestion
    if (savedObject.type === DATA_SOURCE_SAVED_OBJECT_TYPE && (!savedObject.workspaces || savedObject.workspaces.length === 0)) {
```